subsys/bluetooth/host/cs.c

```diff
@@ -360,73 +543,181 @@ void bt_hci_le_cs_subevent_result(struct net_buf *buf)
-	result.header.subevent_done_status = evt->subevent_done_status;
-	result.header.procedure_abort_reason = evt->procedure_abort_reason;
-	result.header.subevent_abort_reason = evt->subevent_abort_reason;
-	result.header.reference_power_level = evt->reference_power_level;
-	result.header.num_antenna_paths = evt->num_antenna_paths;
-	result.header.num_steps_reported = evt->num_steps_reported;
-
-	if (evt->num_steps_reported) {
-		net_buf_simple_init_with_data(&step_data_buf,
-					      evt->steps,
-					      buf->len);
-		result.step_data_buf = &step_data_buf;
+	if (evt->subevent_done_status != BT_HCI_LE_CS_SUBEVENT_DONE_STATUS_PARTIAL) {
+		p_result->step_data_buf = NULL;
+		if (evt->num_steps_reported) {
+			net_buf_simple_init_with_data(&step_data_buf, evt->steps, buf->len);
+			p_result->step_data_buf = &step_data_buf;
+		}
 	} else {
-		result.step_data_buf = NULL;
+		if (evt->procedure_done_status != BT_HCI_LE_CS_PROCEDURE_DONE_STATUS_PARTIAL) {
+			LOG_WRN("Procedure status is inconsistent with subevent status. Discarding "
+				"subevent results");
+			goto abort;
+		}
+
+		if (!evt->num_steps_reported) {
+			LOG_WRN("Discarding partial results without step data");
+			goto abort;
+		}
+
+		reassembly_buf = start_reassembly(conn_handle, evt->steps, buf->len);
+		if (!reassembly_buf) {
+			goto abort;
+		}
+
+		p_result = &reassembled_result;
+		p_result->step_data_buf = &reassembly_buf->b;
+	}
+
+	p_result->header.procedure_counter = sys_le16_to_cpu(evt->procedure_counter);
+	p_result->header.frequency_compensation = sys_le16_to_cpu(evt->frequency_compensation);
+	p_result->header.procedure_done_status = evt->procedure_done_status;
+	p_result->header.subevent_done_status = evt->subevent_done_status;
+	p_result->header.procedure_abort_reason = evt->procedure_abort_reason;
+	p_result->header.subevent_abort_reason = evt->subevent_abort_reason;
+	p_result->header.reference_power_level = evt->reference_power_level;
+	p_result->header.num_antenna_paths = evt->num_antenna_paths;
+	p_result->header.num_steps_reported = evt->num_steps_reported;
+
+	p_result->header.config_id = 0;
+	p_result->header.start_acl_conn_event = 0;
+	if (conn) {
+		p_result->header.config_id = evt->config_id;
+		p_result->header.start_acl_conn_event =
+			sys_le16_to_cpu(evt->start_acl_conn_event_counter);
 	}
 
-#if defined(CONFIG_BT_CHANNEL_SOUNDING_TEST)
-	if (sys_le16_to_cpu(evt->conn_handle) == BT_HCI_LE_CS_TEST_CONN_HANDLE) {
-		result.header.config_id = 0;
-		result.header.start_acl_conn_event = 0;
+	if (evt->subevent_done_status != BT_HCI_LE_CS_SUBEVENT_DONE_STATUS_PARTIAL) {
+		invoke_subevent_result_callback(conn, p_result);
+	}
+
+	if (conn) {
+		bt_conn_unref(conn);
+	}
+
+	if (evt->procedure_done_status != BT_CONN_LE_CS_PROCEDURE_INCOMPLETE) {
+		/* We can now clear the any reassembly buffer allocated for this procedure,
+		 * to avoid code duplication, we're using the abort label to do so
+		 */
+		goto abort;
```

Comment:
if we abort he do we double unref? line 624 and 638?

Reply:
Good catch, need to set it back to NULL

Reply:
or move this branch above line 623?

Reply:
I think as a practice, it's better to set it to NULL when freed.